Add some more information and examples to the image selection documentation (#445)
stackabletech/documentation: modules/concepts/pages/product_image_selection.adoc

These images contain different tools for initialization jobs and/or the actual product itself.
Images are prepared for different architectures and different product versions.

Stackable uses two separate versions to describe the images that are provided as part of the platform:

> **Product Version** +

Reviewer: ```suggestion
**Product version** +
```

> This is the version of the product which this image provides, so this could for example be Kafka 3.3.1

Reviewer: ```suggestion
This is the version of the product which this image provides, so this could for example be Kafka 3.3.1.
```

Reviewer: Maybe we could say "upstream version" as well, to clarify even more that this is not defined by us?

> **Stackable Version** +

Reviewer: ```suggestion
**Stackable version** +
```

This version is used to track changes to the structure of the image containing the product (in the version specified by _product version_).
Stackable operators expect to find a very specific structure in the images they use to roll out the products.
This can be things like startup scripts being present, parameters these startup scripts expect, presence or location of extra libraries and similar things.
In order for our operators to work as intended every operator has to be used with images from the same release line as this operator.

> What this means is, that for example the Stackable Operator for Apache HBase 23.4.0 will by default try to use images with the same stackable version.

Reviewer: ```suggestion
What this means is, that for example the Stackable Operator for Apache HBase 23.4.0 will by default try to use _product_ images with the same Stackable version.
```

I think this sentence can be quite confusing, maybe you could also add an example version number for the product image? Also, how does the operator pick the product version, or is that version mandatory?

Author: That sentence was actually followed by examples originally that I then removed for some reason..

This can be things like startup scripts being present, parameters these startup scripts expect, presence or location of extra libraries and similar things.
In order for our operators to work as intended every operator has to be used with images from the same release line as this operator.

What this means is, that for example the Stackable Operator for Apache HBase will by default try to use images with the same Stackable version, the following table shows a few examples to make this clearer:

|===
|Operator version |HBase version |Image

|23.4.0
|3.3.0
|hbase:3.3.0-stackable23.4.0

|23.4.0
|3.3.1
|hbase:3.3.1-stackable23.4.0

|23.7.0
|3.3.0
|hbase:3.3.0-stackable23.7.0

|23.7.0
|3.3.1
|hbase:3.3.1-stackable23.7.0
|===

However, since the last digit of the Stackable version is considered to be a patchlevel indicator, operators will be compatible with all images from the same release line.
So an operator of version _23.4.x_ will be compatible with all images of version _23.4.y_.
This is intended to allow shorter update cycles for users, when new image versions are released that may contain security fixes - should the user so choose.

> The following paragraphs explain  the available settings and how they work.

Reviewer 2: ```suggestion
The following paragraphs explain the available settings and how they work.
```

Author: done

This affects e.g. configuration properties or available features.
Therefore, the operators are dependent on the product and its product version contained in the custom image.
It's your responsibility to put in the correct product version.

> [WARNING]

Reviewer: I think this block of text could just be part of the text, without the WARNING box. What do you think?

====
Using custom images has a few limitations that users should be aware of.

* The images will *have* to have the same structures that Stackable operators expect.
This should usually be ensured by specifying a Stackable image in the `FROM` clause of the Dockerfile.

* Images will have to be upgraded for every new Stackable release to follow structural changes that Stackable may have made to their images.
When deriving images from official Stackable images this will mean updating the version of the image in the `FROM` clause to the correct Stackable release.

* It is not possible to update the Stackable Platform to a new version without changing the deployed cluster definitions when using custom images.
The recommended process here is:

** Tag clusters as "do not reconcile"
** Update Stackable plattform
** Change custom images in cluster specifications
** Remove "do not reconcile flag"

====

> ## Common Scenarios

Reviewer: ```suggestion
## Common scenarios
```

> ### Planned Plattform Updates

Reviewer: ```suggestion
### Planned platform updates
```

This is probably the most common scenario, users do not specify a Stackable version, and thus the operators always pick the image from their exact release.
Updates happen by updating Stackable Operators, which will in turn restart the products with the new images.

#### Config

[source,yaml]
----
spec:
  image:
    productVersion: 3.3.1
----

> ### Quick Updates of Images

Reviewer: ```suggestion
### Quick updates of images
```

Sometimes it can be useful to decouple operators upgrades from the image versions to allow using updated images as soons as Stackable releases them.
This can significantly shorten turnaround times when reacting to security vulnerabilities for example.

> For this scenario the Stackable Version can be set to the release line, without including the patchlevel indicator.

Reviewer: ```suggestion
For this scenario the Stackable version can be set to the release line, without including the patch level indicator.
```

This will cause the operator to always use the most current image that it is compatible with when starting products.

> [CAUTION]

Reviewer: I think this can be a NOTE, instead of CAUTION, because it says that actually everything will be fine.

Author: done

====
This behavior can result in _mixed_ clusters running on different image versions of the product.
This should not create any issues, since the contained product binaries are exactly the same, but is worth knowing.

A rolling restart of the product would clean this mixed state up.
====

#### Config
[source,yaml]
----
spec:
  image:
    productVersion: 3.3.1
    stackableVersion: 23.4
----

> #### Custom Images / Pinned Images

Reviewer: ```suggestion
#### Custom images / pinned images
```

When a setup requires the utmost stability and it is preferrable for things to break, rather than run with a different image version that for example has not been certified.
Or when a user requires custom libraries / code in the images they run and build their own images derived from official Stackable images, this is the only possible way to do this.

> Please see the warnings in xref:_custom_images for how to upgrade in this scenario.

Reviewer: This xref isn't working

Reviewer: You can also add anchors to section headers which you an then refer to like so:

```
== [[my-section]] My section with a long title

...

please read the <<my-section, section above>>.
```

Author: done

#### Config
[source,yaml]
----
spec:
  image:
    custom: my.corp/myteam/stackable/kafka:latest-and-greatest

>     productVersion: 3.3.1

Reviewer: maybe we could add more explanation here why the productVersion still needs to be specified. But this is optional

Author: The explanation for that is already in the "custom images" section above. We could probably link to it from here if we wanted to ...

This affects e.g. configuration properties or available features.
Therefore, the operators are dependent on the product and its product version contained in the custom image.
It's your responsibility to put in the correct product version.

Using custom images has a few limitations that users should be aware of.

* The images will *have* to have the same structures that Stackable operators expect.
This should usually be ensured by specifying a Stackable image in the `FROM` clause of the Dockerfile.

* Images will have to be upgraded for every new Stackable release to follow structural changes that Stackable may have made to their images.
When deriving images from official Stackable images this will mean updating the version of the image in the `FROM` clause to the correct Stackable release.

* It is not possible to update the Stackable Platform to a new version without changing the deployed cluster definitions when using custom images.
The recommended process here is:

> ** Tag clusters as "do not reconcile"

Reviewer 2: A link to https://docs.stackable.tech/home/stable/concepts/cluster_operations would be great

Author: Ah, right! I made a mental note of that when writing it and then immediately mentally burned the note :)

Author: done
